src/gflownet/tasks/qm9/qm9.py

```diff
@@ -70,21 +74,38 @@ def load_task_models(self):
         gap_model, self.device = self._wrap_model(gap_model)
         return {'mxmnet_gap': gap_model}
 
-    def sample_conditional_information(self, n):
+    def sample_conditional_information(self, n: int) -> Dict[str, Tensor]:
         beta = None
-        if self.temperature_sample_dist == 'gamma':
-            beta = self.rng.gamma(*self.temperature_dist_params, n).astype(np.float32)
-        elif self.temperature_sample_dist == 'uniform':
-            beta = self.rng.uniform(*self.temperature_dist_params, n).astype(np.float32)
-        elif self.temperature_sample_dist == 'beta':
-            beta = self.rng.beta(*self.temperature_dist_params, n).astype(np.float32)
-        beta_enc = thermometer(torch.tensor(beta), 32, 0, 32)  # TODO: hyperparameters
-        return {'beta': torch.tensor(beta), 'encoding': beta_enc}
-
-    def cond_info_to_reward(self, cond_info: Dict[str, Tensor], flat_reward: FlatRewards) -> RewardScalar:
+        if self.temperature_sample_dist == 'constant':
+            assert type(self.temperature_dist_params) in [float, int]
+            beta = np.array(self.temperature_dist_params).repeat(n).astype(np.float32)
+            beta_enc = torch.zeros((n, self.num_thermometer_dim))
+        else:
+            if self.temperature_sample_dist == 'gamma':
+                loc, scale = self.temperature_dist_params
+                beta = self.rng.gamma(loc, scale, n).astype(np.float32)
+                upper_bound = stats.gamma.ppf(0.95, loc, scale=scale)
+            elif self.temperature_sample_dist == 'uniform':
+                beta = self.rng.uniform(*self.temperature_dist_params, n).astype(np.float32)
+                upper_bound = self.temperature_dist_params[1]
+            elif self.temperature_sample_dist == 'loguniform':
+                beta = np.exp(self.rng.uniform(*np.log(self.temperature_dist_params), n).astype(np.float32))
+                upper_bound = self.temperature_dist_params[1]
+            elif self.temperature_sample_dist == 'beta':
+                beta = self.rng.beta(*self.temperature_dist_params, n).astype(np.float32)
+                upper_bound = 1
+            beta_enc = thermometer(torch.tensor(beta), self.num_thermometer_dim, 0, upper_bound)
+
+        assert len(beta.shape) == 1, f"beta should be a 1D array, got {beta.shape}"
+        return {'beta': beta, 'encoding': beta_enc}
+
+    def cond_info_to_logreward(self, cond_info: Dict[str, Tensor], flat_reward: FlatRewards) -> RewardScalar:
         if isinstance(flat_reward, list):
             flat_reward = torch.tensor(flat_reward)
-        return RewardScalar(flat_reward.flatten()**cond_info['beta'])
+        flat_reward = flat_reward.squeeze()
+        assert len(flat_reward.shape) == len(cond_info['beta'].shape), \
+            f"dangerous shape mismatch: {flat_reward.shape} vs {cond_info['beta'].shape}"
+        return RewardScalar(torch.log(flat_reward + 1e-8) * cond_info['beta'])
```

Comment:
I'd do a `.clamp` here, since it doesn't distort regular values and is not prone to NaN when going through gradients (which shouldn't be the case here, but better to use that convention throughout the code).

Reply:
Were you thinking about something like:
```
return RewardScalar(torch.max(torch.log(flat_reward), torch.tensor([-75.])) * cond_info['beta'])
```

Reply:
It's safer to clamp before `log` (for 0, subnormal floats, or negative values), e.g: 

https://github.com/recursionpharma/gflownet/blob/450098e371ea2eb71c37f1f3acf64f8e65149036/src/gflownet/algo/trajectory_balance.py#L263

Reply:
[Fixed here](https://github.com/recursionpharma/gflownet/commit/e3fdb25548395fce6598f763e391ef48d123a3c6)